metricbeat/module/system/services/data.go

```diff
@@ -0,0 +1,171 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package services
+
+import (
+	"time"
+
+	"github.com/coreos/go-systemd/dbus"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+// formProperties gets properties for the systemd service and returns a MapStr with useful data
+func formProperties(unit dbus.UnitStatus, props map[string]interface{}) (common.MapStr, error) {
+	timeSince, err := timeSince(props, unit.ActiveState)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting timestamp")
+	}
+
+	event := common.MapStr{
+		"name":         unit.Name,
+		"load_state":   unit.LoadState,
+		"active_state": unit.ActiveState,
+		"sub_state":    unit.SubState,
+	}
+
+	//most of the properties values are context-dependent.
+	//If things aren't running/haven't run/etc than a lot of the values should be ignored.
+
+	//Even systemd doesn't check the substate, leading to a lot of odd `Memory: 0B` lines in `systemctl status`
+	//Ignore the resource accounting if a service has exited
+	if unit.ActiveState == "active" && unit.SubState != "exited" {
+		event["resources"] = getMetricsFromServivce(props)
+
+	}
+
+	//anything less than 1 isn't a valid SIGCHLD code
+	if props["ExecMainCode"].(int32) > 0 {
+		event["exec_code"] = translateChild(props["ExecMainCode"].(int32))
+		event["exec_rc"] = props["ExecMainStatus"].(int32)
```

Comment:
This kind of type conversation is always scarry. This can panic a Beat if something changes and is not hit be a test. Is it possible to work with concrete types when collecting the properties and postpone the conversation to a dynamic type?